ash/src/lib.rs

```diff
@@ -66,6 +66,40 @@ impl<'r, T> RawPtr<T> for Option<&'r T> {
     }
 }
 
+/// Given a mutable raw pointer to a type with an `s_type` member such as [`vk::BaseOutStructure`],
+/// match on a set of Vulkan structures. The struct will be rebound to the
+/// given variable of the type of the given Vulkan structure.
+///
+/// ```
+/// let mut info = ash::vk::DeviceCreateInfo::default();
+/// let info: *mut ash::vk::BaseOutStructure = <*mut _>::cast(&mut info);
+/// unsafe {
+///     ash::match_struct!(match info {
+///         info @ ash::vk::DeviceQueueCreateInfo => {
+///             dbg!(&info); // Unreachable
+///         }
+///         info @ ash::vk::DeviceCreateInfo => {
+///             dbg!(&info);
+///         }
+///     })
+/// }
+/// ```
+#[macro_export]
+macro_rules! match_struct {
+    (match $p:ident { $($bind:ident @ $ty:path => $body:block $(,)?)+ }) => {
+        match (*$p).s_type {
+            $(<$ty as $crate::vk::HasStructureType>::STRUCTURE_TYPE => {
+                let $bind = $p
+                    .cast::<$ty>()
+                    .as_mut()
+                    .unwrap();
+                $body
+            }),+
+            _ => {}
```

Comment:
This catch-all is currently preventing the use of this `match_struct` to return values. Perhaps we can employ some sort of optional `$(_ => $...)?` argument to allow (or if necessary, force) the caller to set something here? I'd expect them to mostly use an empty block `{}` or `()` void, or `None` when otherwise using the macro to look for a specific type.

Reply:
`Option` seems sensible; alternatively, the macro could require the caller to explicitly provide a default case. Or we could do nothing; after all, people can always assign via side effects if they want.

Reply:
There's now an optional `_` wildcard arm for this case, where users can write their own body/expression just like the other match arm. Including a tested example in the docs.

Reply:
Nice, that's consistent with how `if` can have a non-unit value iff you have an `else` branch then.

Reply:
It's the same for any normal `match` block as well, considering all Vulkan enums should be `non_exhaustive` most definitely including `StructureType`. We don't annotate ours as such currently :sweat_smile:

Reply:
That's because our enums aren't actually enums, they're all just newtyped integers with associated constants. `#[non_exhaustive]` enums wouldn't even work here, since it's still UB to transmute an unused bitpattern into those.

Reply:
Nvm, I got this mixed up with some other crate that tried to do this. These integers are nonexhaustive anyways (unless you actually match on every possible value/range, I guess?)